crates/shared/src/price_estimation/competition.rs

```diff
@@ -34,62 +28,53 @@ impl RacingCompetitionPriceEstimator {
             successful_results_for_early_return,
         }
     }
+}
 
-    fn enough_successful_estimates_for_each_query(
-        &self,
-        queries: &[Query],
-        results: &[(&String, Vec<PriceEstimateResult>)],
-    ) -> bool {
-        for i in 0..queries.len() {
-            if results.iter().filter(|result| result.1[i].is_ok()).count()
-                < self.successful_results_for_early_return.into()
-            {
-                return false;
-            }
-        }
-
-        true
-    }
-
-    async fn estimates_<'a>(
+impl PriceEstimating for RacingCompetitionPriceEstimator {
+    fn estimates<'a>(
         &'a self,
         queries: &'a [Query],
-    ) -> impl Iterator<Item = PriceEstimateResult> + 'a {
+    ) -> futures::stream::BoxStream<'_, (usize, PriceEstimateResult)> {
         debug_assert!(queries.iter().all(|query| {
             query.buy_token != model::order::BUY_ETH_ADDRESS
                 && query.sell_token != model::order::BUY_ETH_ADDRESS
                 && query.sell_token != query.buy_token
         }));
 
-        let mut remaining_estimates: Vec<_> = self
-            .inner
-            .iter()
-            .map(|(name, estimator)| {
-                async move { (name, vec_estimates(estimator.as_ref(), queries).await) }.boxed()
-            })
-            .collect();
-
-        let mut computed_estimates = Vec::with_capacity(queries.len());
-
-        while !remaining_estimates.is_empty() {
-            let (estimate, _, rest) = future::select_all(remaining_estimates).await;
-            computed_estimates.push(estimate);
-            if self.enough_successful_estimates_for_each_query(queries, &computed_estimates) {
-                break;
+        let combined_stream = futures::stream::select_all(self.inner.iter().enumerate().map(
+            |(i, (_, estimator))| estimator.estimates(queries).map(move |result| (i, result)),
+        ));
```

Comment:
nit: I find it a little confusing that the code to handle the `combined_streams` is split in two parts with `handle_single_result` in between. AFAICS this part could as well move to the bottom where `combined_stream` gets used to not have related logic together.

Reply:
The reason this code isn't directly in filter_map below is that I wanted to avoid more rightward drift as it would be indented an extra level and that I wanted to be able to use `?` inside of it which I can't do if it's inline. Then there was some extra annoyances where I couldn't just put an async block down there because it would run into lifetime problems.

The reason I ordered it this way is it made sense to me that you first see how the streams are combined and then how each result in the combined stream is handled. The filter_map really just glues them together but doesn't do anything interesting.

Reply:
I totally agree that it's worth having a named closure for `handle_single_result`.

I can follow your reasoning for "interleaving" the code I just would have it done differently which added a bit of cognitive load to `estimates()` for me. I thought I would have to keep all the code of the `combined_stream` logic and `handle_single_result()` in memory at the same time for me to understand what's going on.

With the comments added to `handle_single_result()` it's easier to understand separate chunks of the logic so I don't feel strongly about reordering the logic anymore.